File: src/VisualBasic/CodeCracker/Performance/RemoveWhereWhenItIsPossibleAnalyzer.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports System.Collections.Immutable
Imports CodeCracker.Extensions

Namespace Performance
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class RemoveWhereWhenItIsPossibleAnalyzer
        Inherits DiagnosticAnalyzer

        Public Shared ReadOnly Id As String = DiagnosticId.RemoveWhereWhenItIsPossible.ToDiagnosticId()
        Public Const Title As String = "You should remove the 'Where' invocation when it is possible."
        Public Const MessageFormat As String = "You can remove 'Where' moving the predicate to '{0}'."
        Public Const Category As String = SupportedCategories.Performance
        Public Const Description As String = "When a LINQ operator supports a predicate parameter it should be used instead of using 'Where' followed by the operator"
        Protected Shared Rule As DiagnosticDescriptor = New DiagnosticDescriptor(
            Id,
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Warning,
            isEnabledByDefault:=True,
            description:=Description,
            helpLink:=HelpLink.ForDiagnostic(Id))

        Public Overrides ReadOnly Property SupportedDiagnostics() As ImmutableArray(Of DiagnosticDescriptor) = ImmutableArray.Create(Rule)

        Shared ReadOnly supportedMethods() As String = {"First", "FirstOrDefault", "Last", "LastOrDefault", "Any", "Single", "SingleOrDefault", "Count"}

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf AnalyzeNode, SyntaxKind.InvocationExpression)
        End Sub

        Private Sub AnalyzeNode(context As SyntaxNodeAnalysisContext)
            Dim whereInvoke = DirectCast(context.Node, InvocationExpressionSyntax)
            If GetNameOfTheInvokeMethod(whereInvoke) <> "Where" Then Exit Sub

            Dim nextMethodInvoke = whereInvoke.Parent.FirstAncestorOrSelf(Of InvocationExpressionSyntax)()

            Dim candidate = GetNameOfTheInvokeMethod(nextMethodInvoke)
            If Not supportedMethods.Contains(candidate) Then Exit Sub

            If nextMethodInvoke.ArgumentList.Arguments.Any Then Return

            Dim diag = Diagnostic.Create(Rule, GetNameExpressionOfTheInvokedMethod(whereInvoke).GetLocation(), candidate)

            context.ReportDiagnostic(diag)
        End Sub

        Friend Shared Function GetNameOfTheInvokeMethod(invoke As InvocationExpressionSyntax) As String
            If (invoke Is Nothing) Then Return Nothing
            Dim memberAccess = invoke.ChildNodes.
            OfType(Of MemberAccessExpressionSyntax).
            FirstOrDefault()

            Return GetNameExpressionOfTheInvokedMethod(invoke)?.ToString()
        End Function

        Friend Shared Function GetNameExpressionOfTheInvokedMethod(invoke As InvocationExpressionSyntax) As SimpleNameSyntax
            If invoke Is Nothing Then Return Nothing

            Dim memberAccess = invoke.ChildNodes.
            OfType(Of MemberAccessExpressionSyntax)().
            FirstOrDefault()

            Return memberAccess?.Name

        End Function

    End Class
End Namespace

File: src/CSharp/CodeCracker/Performance/RemoveWhereWhenItIsPossibleAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace CodeCracker.CSharp.Performance
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class RemoveWhereWhenItIsPossibleAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "You should remove the 'Where' invocation when it is possible.";
        internal const string MessageFormat = "You can remove 'Where' moving the predicate to '{0}'.";
        internal const string Category = SupportedCategories.Performance;
        const string Description = "When a linq operator support a predicate parameter it should be used instead of "
            + "using 'Where' followed by the operator";

        static readonly string[] supportedMethods = new[] {
            "First",
            "FirstOrDefault",
            "Last",
            "LastOrDefault",
            "Any",
            "Single",
            "SingleOrDefault",
            "Count"
        };

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.RemoveWhereWhenItIsPossible.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Warning,
            isEnabledByDefault: true,
            description: Description,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.RemoveWhereWhenItIsPossible));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) =>
            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.InvocationExpression);

        private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var whereInvoke = (InvocationExpressionSyntax)context.Node;
            var nameOfWhereInvoke = GetNameOfTheInvokedMethod(whereInvoke);
            if (nameOfWhereInvoke?.ToString() != "Where") return;
            if (ArgumentsDoNotMatch(whereInvoke)) return;

            var nextMethodInvoke = whereInvoke.Parent.
                FirstAncestorOrSelf<InvocationExpressionSyntax>();
            if (nextMethodInvoke == null) return;

            var candidate = GetNameOfTheInvokedMethod(nextMethodInvoke)?.ToString();
            if (!supportedMethods.Contains(candidate)) return;

            if (nextMethodInvoke.ArgumentList.Arguments.Any()) return;
            var properties = new Dictionary<string, string> { { "methodName", candidate } }.ToImmutableDictionary();
            var diagnostic = Diagnostic.Create(Rule, nameOfWhereInvoke.GetLocation(), properties, candidate);
            context.ReportDiagnostic(diagnostic);
        }

        private static bool ArgumentsDoNotMatch(InvocationExpressionSyntax whereInvoke)
        {
            var arguments = whereInvoke.ArgumentList.Arguments;
            if (arguments.Count != 1) return true;
            var expression = arguments.First()?.Expression;
            if (expression == null) return true;
            if (expression is SimpleLambdaExpressionSyntax) return false;
            var parenthesizedLambda = expression as ParenthesizedLambdaExpressionSyntax;
            if (parenthesizedLambda == null) return true;
            return parenthesizedLambda.ParameterList.Parameters.Count != 1;
        }

        private static SimpleNameSyntax GetNameOfTheInvokedMethod(InvocationExpressionSyntax invoke) =>
            invoke.ChildNodes().OfType<MemberAccessExpressionSyntax>().FirstOrDefault()?.Name;
    }
}